Foundation/NSString.swift

```diff
@@ -1297,38 +1297,137 @@ extension NSString {
+        let cfEncodings = CFStringGetListOfAvailableEncodings()
+        var encodingPtr = cfEncodings
+        
+        while encodingPtr?.pointee != kCFStringEncodingInvalidId {
+            if let cfEncodingName = CFStringConvertEncodingToIANACharSetName(encodingPtr!.pointee) {
+                var encodingName: String?
+                if String._conditionallyBridgeFromObjectiveC(cfEncodingName._nsObject, result: &encodingName) {
+                    if encodingName == encodingStr {
+                        let encoding = CFStringConvertEncodingToNSStringEncoding(encodingPtr!.pointee)
+                        return String.Encoding.init(rawValue: encoding)
+                    }
+                } else {
+                    continue
+                }
+            }
+            
+            encodingPtr = encodingPtr?.advanced(by: 1)
+        }
+        
+        return nil
+    }
+    
     public convenience init(contentsOfFile path: String, usedEncoding enc: UnsafeMutablePointer<UInt>?) throws {
-        NSUnimplemented()    
+        let readResult = try Data(contentsOf: URL(fileURLWithPath: path), options:[])
+        var encoding: String.Encoding?
+        
+        // Check extended attributes for 'com.apple.TextEncoding'
+        let attrName = "com.apple.TextEncoding"
+        let bufCount = getxattr(path, attrName, nil, 0, 0, 0)
+        if bufCount > 0 {
+            var buf = [UInt8](repeating: 0, count: bufCount)
+            if getxattr(path, attrName, &buf, bufCount, 0, 0) != -1 {
+                if let attrValue = String(bytes: buf, encoding: .utf8) {
+                    encoding = NSString._getEncodingNameFromString(attrValue)
+                }
+            }
+        }
+        
+        // If the encoding can't be found in extended attributes, check for a BOM
+        if encoding == nil {
+            // If the encoding can't be found, use utf8 as the default
+            encoding = NSString._getEncodingFromDataByCheckingForUnicodeBOM(readResult) ?? .utf8
+        }
+        
+        enc?.pointee = encoding!.rawValue
+        
+        guard let cf = NSString._createCFString(fromData: readResult, withEncoding: encoding!) else {
+            throw NSError(domain: NSCocoaErrorDomain,
+                          code: CocoaError.fileReadUnknownStringEncoding.rawValue,
+                          userInfo: [
+                            "NSDebugDescription" : "The file \"\(path)\" couldn't be opened because the text encoding of its contents can't be determined.",
+                            "NSFilePath": path
+                ])
+        }
+        
+        var str: String?
+        if String._conditionallyBridgeFromObjectiveC(cf._nsObject, result: &str) {
+            self.init(str!)
+        } else {
+            throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.fileReadUnknownStringEncoding.rawValue, userInfo: [
```

Comment:
Which one, the bridging? I was going by other code that handled the same thing- I can convert to a fatal error if that's preferred.